bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -171,22 +225,27 @@ void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequest
       LOG_INFO(GL, "--preProcessor metrics dump--" + metricsComponent_.ToJson());
     }
   }
+}
+
+template <>
+void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequestMsg *msg) {
+  updateAggregatorAndDumpMetrics();
   preProcessorMetrics_.preProcReqReceived.Get().Inc();
   ClientPreProcessReqMsgUniquePtr clientPreProcessReqMsg(msg);
 
   MDC_CID_PUT(GL, clientPreProcessReqMsg->getCid());
   const NodeIdType &senderId = clientPreProcessReqMsg->senderId();
   const NodeIdType &clientId = clientPreProcessReqMsg->clientProxyId();
   const ReqId &reqSeqNum = clientPreProcessReqMsg->requestSeqNum();
-  LOG_INFO(GL,
-           "Received ClientPreProcessRequestMsg reqSeqNum=" << reqSeqNum << " from clientId=" << clientId
-                                                            << ", senderId=" << senderId);
+  LOG_DEBUG(GL,
+            "Received ClientPreProcessRequestMsg reqSeqNum=" << reqSeqNum << " from clientId=" << clientId
+                                                             << ", senderId=" << senderId);
   if (!checkClientMsgCorrectness(clientPreProcessReqMsg, reqSeqNum)) {
     preProcessorMetrics_.preProcReqIgnored.Get().Inc();
     return;
   }
   {
-    auto &clientEntry = ongoingRequests_[clientId];
+    const auto &clientEntry = ongoingRequests_[clientId];
     lock_guard<mutex> lock(clientEntry->mutex);
```

Comment:
Let's set the lock first:
```c++
    lock_guard<mutex> lock(clientEntry->mutex);
    const auto &clientEntry = ongoingRequests_[clientId];
```

Reply:
That's all the idea - clientEntry (ClientRequestInfoSharedPtr) always exists. What is changed is RequestProcessingInfoUniquePtr. That's why we can safely get ongoingRequests_[clientId], but then it should be locked to get  RequestProcessingInfo.
Actually I think you were right and RequestProcessingInfo should be renamed to RequestProcessingState to be more understandable.

Reply:
Thanks Yulia, I appreciate the improvement!